File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
          
            
        List<List<Integer>> ans = new ArrayList<>();
         if(root==null) return ans;
        Queue<TreeNode> q1 = new ArrayDeque<>();
            boolean flag=false;  
            q1.offer(root);
       
            while(!q1.isEmpty()){                                                           
                     
                       
                              int n = q1.size();  
                          List<Integer> list = new ArrayList<>();
                         for(int i=0; i<n; i++){
                            TreeNode temp = q1.poll();
                                  if(temp.left!=null)
                                    q1.offer(temp.left);
                            if(temp.right!=null)
                                   q1.offer(temp.right);                           
                            list.add(temp.val); 
                             
                          }
                if(flag){
                    Collections.reverse(list);
                }
                flag = !flag;
                 ans.add(list);                                
                   
    
                }
        return ans;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q = []
        if root:
            q.append(root)
        res = []
        flag = False
        while q:
            val = []
            
            for i in range(len(q)):
                temp = q.pop(0)
                if temp.left:
                    q.append(temp.left)
                if temp.right:
                    q.append(temp.right)
                val.append(temp.val)
            if flag:
                val.reverse()
            flag = not flag
            res.append(val)
        
        return res
        